Камень судьбы

AlexBad opened the pull request:
---
:mortar_board: [Камень судьбы](https://up.htmlacademy.ru/yii/1/user/1540437/tasks/3)

AlexBad committed:
Предварительный вариант

AlexBad committed:
fix

Commit:
fix 2

AlexBad committed:
Задача 2

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+ * @property int $performer_id ID Исполнителя 
+ * @property int $customer_id ID Заказчика
+ * @property string $status Статус задачи
+ * @property string $role Роль пользваотеля
```
Слово "пользователя" пьяненькое =)

AlexBad replied:
Исправлю

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+    /** Исполнитель начинает выполнение задания */
+    const ACTION_PERFORMER_START = 'actionStart';
+
+    protected static $_constants;
```
Фишка с $_constants чрезмерная автоматизация и усложнение.
Лучше сделать проще, для каждого типа констант объявить переменную-массив и перечислить их все там.

int64 replied:
Проще говоря:
```php
const ACTIONS = [self::ACTION_PERFORMER_START, self::ACTION...];
```

AlexBad replied:
Хорошо.

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+    protected static $_constants;
+
+    public $id;
+    public $customer_id;
```
Здесь и везде следует выбрать правило именования camelCase или snake_case. Рекомендую camelCase. Это касается и свойств и переменных и методов.

AlexBad replied:
camelCase

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+    /**
+     * Создание новой схемы 'Задачи'.
+     * 
+     * @param integer $customer ID Заказчика
```
Также стоит определиться с типами. Тут integer, в определении класса int. Советую в doc-блоках описывать тот же тип что и будет непосредственно в коде, в данном случае int.
Подробнее про типы аргументов можно почитать тут https://www.php.net/manual/ru/functions.arguments.php#functions.arguments.type-declaration.

AlexBad replied:
Не вижу причин не последовать твоему совету :)

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+     * @param integer $customer ID Заказчика
+     * @param integer $performer ID Исполнителя
+     */
+    public function __construct($customer = null, $performer = null)
```
Если версия PHP позволяет, то стараемся использовать типизацию аргументов, возвращаемых значений и всего, что можно =). Начиная с 7.4 можно объявлять типы свойств.

AlexBad replied:
Поправлю

AlexBad replied:
На домашнем компьютере стоит PHP 7.2... не пропускает типизацию свойств.

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+     */
+    public function __construct($customer = null, $performer = null)
+    {
+        $this->id = random_int(1, 999);
+        
+        if (is_null($customer)) {
+            $customer = random_int(1,30);
+        }
+        if (is_null($performer)) {
+            $performer = random_int(31, 80);
+        }
```
Это неправильное поведение класса, нельзя насильно определять значения свойств, если того не требует бизнес-логика, да еще и рандомно. В общем, нужно просто убрать.

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+    /** Заказчик разместил новое задание */
+    public function actionCreate()
+    {
+        $this->id = random_int(10, 9999);
```
Тоже самое

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+    const STATUS_FAIL = 'fail';
+
+    /** Заказчик разместил новое задание */
+    const ACTION_CUSTOMER_CREATE = 'actionCreate';
```
Префикс action из значения нужно убрать.

int64 replied:
На схеме https://up.htmlacademy.ru/yii/1/project/task-force#taskforce-5-1
только 4 действия, а у нас получилось 7.

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+    }
+
+    /** Заказчик отменил задание */
+    public function actionCancel()
+    {
+        if ($this->canRunAction(__FUNCTION__) && $this->changeStatus(self::STATUS_CANCELED)) {
+            return $this->status;
+        }        
+        return false;
+        
+    }
```
1. Код методов action одинаковый, можно вынести в отдельный метод.
2. В метод canRunAction передается имя метода, а нужно как раз использовать константы.

AlexBad replied:
Пересмотрел реализацию всех action методов.

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+    /** Исполнитель предложил свои услуги */
+    public function actionPending()
+    {
+        if ($this->canRunAction(__FUNCTION__)) {
```
Тут нет изменения статуса

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+    private function changeStatus(string $status)
+    {
+        if (!isset($this->status)) {
+            $this->status = self::STATUS_NEW;
```
Вот это как раз лучше делать в конструкторе.

AlexBad replied:
Перенес в конструктор

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+        if (in_array($status, $statusList)) {
+            return true;
+        } else {
+            echo "Состояние '{$status}' не входит с писок допустимых.\n";
```
Пока уберем вывод ошибок, вернемся к ним когда будем использовать исключения.
Хотя в данном случае ошибок быть не должно, метод просто должен вернуть валидный статус или нет.

AlexBad replied:
О, я это сделал для само проверки... Видимо нужно чистить код перед заливкой.

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+     */
+    private static function isActionValid($action)
+    {
+        return in_array($action, self::ActionList()) ? true : false;
```
Т.к. in_array итак возвращает bool, то можно просто сделать так:
```suggestion
        return in_array($action, self::ActionList());
```

AlexBad replied:
Убрал лишнее

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,315 @@
+     * @param string $prefix Prefix
+     * @return array[] $constants
+     */
+    private static function constants(string $prefix = '')
```
Этот метод надо убить =).

AlexBad replied:
Убил

AlexBad committed:
Исправления по коментариям

Commit:
Исправления с учетом комментариев.
Изменил фаил index.php для само проверки.

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,249 @@
+     */
+    public function __construct(int $customer, ?int $performer = null)
+    {
+        if (!isset($this->status)) {
```
Это лишнее, статус при создании всегда будет null

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,249 @@
+     * @throws NotAllowedActionException Если не возможно выполнить `Действие`.
+     * @return void    
+     */
+    public function changeStatus(string $status, string $action): void
```
По логике мы не можем снаружи менять статус, только внутри класса, потому что статус это состояние.
Т.е. логика должна быть такой: снаружи пытаемся выполнить какое-то действие, а оно (действие) уже внутри меняет статус.
Плюс неплохо было бы разделить ответственность между разными методами. Например, один проверяет действие, второй проверяет возможность перейти в статус, третий уже выполняет действие.
Также посмотри доступность методов, далеко не все методы должны быть публичными. Скажем так, по умолчанию метод приватный, а дальше мы смотрим насколько его должно быть видно.

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,249 @@
+}
+
+
+class NotAllowedActionException extends Exception
```
Свои эксепшены это хорошо, молодец.
Единственное, добавь еще базовый класс своих эксепшенов и эти уже наследуй от него, так в будущем будет удобнее работать с ними.

AlexBad committed:
Предварительный вариант

Commit:
Версия #3 на проверку - Task.php

Commit:
И снова правки %)

int64 commented on src/Task.php:
```diff
@@ -0,0 +1,376 @@
+     * 
+     * @return array
+     */
+    private function getStatusList(): array
+    {
+        $map = [];
+        $map[self::STATUS_NEW][] = self::STATUS_CANCELED;
+
+        if ($this->isHasPerformer()) {
+            $map[self::STATUS_NEW][] = self::STATUS_INPROGRESS;
+            $map[self::STATUS_INPROGRESS] = [
+                self::ACTION_PERFORMER_REFUSE,
+                self::STATUS_COMPLETE,
+                self::STATUS_FAIL
+            ];
+        }
+
+        return $map;
+    }
```
1. Не совсем понял, что возвращает этот метод. Где-то в значениях статусы, где-то - действия
2. Так же не понял почему описаны статусы только для исполнителя. Лучше это вынести в отдельное место. Т.е. из какого статуса в какой описывается в одном месте, а для какого пользователя какие действия доступны в другом.

AlexBad replied:
1.  ```self::ACTION_PERFORMER_REFUSE``` - Лишний :) Тут должны быть только Состояния. Уберу.
Метод должен возвращать список доступных переходов Состояния. Тоесть, я заранее знаю, какие Состояние возможны, в текущем Состоянии.
2. Состояния описаны для Всех. Тут проверка, на наличие "Исполнителя" в задании. "Карта" доступных переходов, сильно меньше, если нет Исполнителя.

Думаешь, нужно делать иначе?

int64 replied:
1. Ок
2. В данном месте мы описываем только логику доступных действий. Самих действий при этом меньше не становится, просто некоторые действия доступны только для определенной роли и это мы как раз вынесем в другое место. Т.е. при попытке выполнить какое-то действие мы в будущем будем делать две проверки, одна про последовательность действий, другая про доступность действия текущему пользователю.

Commit:
Версия 4, на проверку Task.php

AlexBad committed:
Убрали логику работы с Пользователяит и ролями.
Добавил обязательный параметр для Выполнения действия - string $role
Осталась Карта Действий в зависимости от Роли.
